Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClientCache.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using BuildXL.Cache.ContentStore.Exceptions;
+using BuildXL.Cache.ContentStore.Interfaces.Results;
+using BuildXL.Cache.ContentStore.Interfaces.Tracing;
+
+namespace BuildXL.Cache.ContentStore.Service.Grpc
+{
+    /// <summary>
+    /// Cache for <see cref="GrpcCopyClient"/>.
+    /// </summary>
+    public sealed class GrpcCopyClientCache
+    {
+        // Link to DistributedContentStoreSettings.MaxConcurrentCopyOperations
+        internal readonly ConcurrentDictionary<GrpcCopyClientKey, GrpcCopyClient> _clientDict;
+
+        private readonly object _backgroundCleanupLock = new object();
+        private readonly int _maxClientCount;
+        private readonly int _maximumAgeInMinutes;
+        private readonly int _waitBetweenCleanupInMinutes;
+        private Context _context;
+        private Task _backgroundCleaningTask;
+        private CancellationTokenSource _backgroundCleaningTaskTokenSource;
+
+        /// <summary>
+        /// Cache for <see cref="GrpcCopyClient"/>.
+        /// </summary>
+        /// <param name="context">Content.</param>
+        /// <param name="maxClientCount">Maximum number of clients to cache.</param>
+        /// <param name="maxClientAgeMinutes">Maximum age of cached clients.</param>
+        /// <param name="waitBetweenCleanupMinutes">Minutes to wait between cache purges.</param>
+        public GrpcCopyClientCache(Context context, int maxClientCount = 512, int maxClientAgeMinutes = 55, int waitBetweenCleanupMinutes = 17)
+        {
+            _context = context;
+            _maximumAgeInMinutes = maxClientAgeMinutes;
+            _waitBetweenCleanupInMinutes = waitBetweenCleanupMinutes;
+            _maxClientCount = maxClientCount;
+
+            _clientDict = new ConcurrentDictionary<GrpcCopyClientKey, GrpcCopyClient>(Environment.ProcessorCount, _maxClientCount);
+            StartBackgroundCleanup();
+        }
+
+        internal void StartBackgroundCleanup()

```

Comment:
I'm not sure, but the fact that this method internal and exposed made a lot of stuff way more complicated.

First of all, I can't see why we need to expose this method at all...

Reply:
For testing, really. I didn't want to write a test which relies on cleanup without having a way to initiate cleanup on demand. I don't want to keep this cleanup feature without having tests to validate it.

---
In reply to: [290519565](https://github.com/microsoft/BuildXL/pull/391#discussion_r290519565) [](ancestors = 290519565)